### PR TITLE
Sum

### PR DESCRIPTION
**Pull request by AlexKnauth**
This uses `Sum` for `Operation`, and a combination of `Sum` and `delay-type` for `TransactionStatus`

### CODE REVIEW

File: json-rpc.ss [hunk 1]
```diff
@@ -104,9 +104,9 @@
 
 (def ToData<-Operation
   (match <>
-    ((TransferTokens recipient) (values recipient (void)))
-    ((CreateContract code) (values (void) code))
-    ((CallFunction recipient data) (values recipient data))))
+    ((Operation-TransferTokens recipient) (values recipient (void)))
+    ((Operation-CreateContract code) (values (void) code))
+    ((Operation-CallFunction {to: recipient (data)}) (values recipient data))))
```
Comment (fare) [inline json-rpc.ss:109]: Make the last case `{(to) (data)}`

File: tx-tracker.ss [hunk 2]
```diff
@@ -250,15 +210,15 @@
...
+               ((TransactionStatus-TxWanted pretx)
                 (match (with-result (make-signed-transaction pretx))
                   ((failure e) (invalidate status e))
-                  ((some stx) (continue (TxSigned pretx stx)))))
-               ((TxSigned pretx signed)
+                  ((some stx) (continue (TransactionStatus-TxSigned (vector pretx stx))))))
+               ((TransactionStatus-TxSigned (vector pretx signed))
                 (match (with-result
-                        (retry retry-window: 0.05 max-window: 30.0 max-retries: +inf.0
+                        (retry retry-window: 0.05 max-window: 30.0 max-retries: 10 ; +inf.0
```
Comment (fare) [inline tx-tracker.ss:221]: Restore the `+inf.0` ?

File: types.ss [hunk 3]
```diff
@@ -16,14 +16,30 @@
...
-  (prefix-in (only-in :clan/poo/number Nat UInt. IntSet) poo.)
+  (prefix-in (only-in :clan/poo/number Nat UInt. UInt IntSet) poo.)
   (prefix-in :clan/poo/type poo.)
   :clan/poo/brace
   ./hex ./abi)
 
 ;; --- something for types in general, including Record, Union, Maybe
 ;; --- something for ethereum types in particular
 
+(.def (DelayedType @ [Type.] sexp .get-delegate)
```
Comment (fare) [inline types.ss:27]: You're not using `sexp`, are you? Actually, you should probably define it in a deferred way, too.

File: types.ss [hunk 4]
```diff
@@ -287,6 +303,64 @@
     (.<-nat (u8vector-uint-ref bytes (+ head .ethabi-padding) big .length-in-bytes))))
 (defrule (Enum values ...) {(:: @ Enum.) vals: '(values ...)})
 
+;; Sum : {Kw Type} ... -> Type
+(def (Sum . plist)
+  ;; a : [Assocof Symbol Type]
+  (def a (map (match <> ([kw . type] (cons (symbolify kw) type))) (alist<-plist plist)))
+  (def tag-marsh-t (poo.UInt (integer-length (max 0 (1- (length a))))))
```
Comment (fare) [inline types.ss:310]: Why have an integer at all for 0 cases? or 1 case?

File: types.ss [hunk 5]
```diff
@@ -287,6 +303,64 @@
...
+  (def t
+    {(:: @ [methods.bytes<-marshal Type.])
+      sexp: ['Sum . plist]
```
Comment (fare) [inline types.ss:313]: I don't see the point of defining `t` that is only used once, to return it immediately. Just define it inline. Use `@` as the open-ended self-reference, if needed (doesn't seem so).

File: tx-tracker.ss (hunk 2 above)
Comment (AlexKnauth) [reply tx-tracker.ss:221]: Oh, thanks for catching that, I'll fix it

File: types.ss (hunk 3 above)
Comment (AlexKnauth) [reply types.ss:27]: The `delay-type` macro defines the `sexp` slot when it invokes `DelayedType`. But I suppose `DelayedType` itself doesn't need to reference that

File: types.ss (hunk 4 above)
Comment (AlexKnauth) [reply types.ss:310]: For 0 cases it doesn't matter, there are no values. For 1 case, would you prefer `(Sum tag: Type)` to return `Type`?

Comment (AlexKnauth) [reply types.ss:311]: We're leaving `(Sum tag: Type)` different from `Type` for now, which is consistent with `(Tuple Type)` and `(Record field: Type)` also wrapping them.

File: types.ss (hunk 5 above)
Comment (AlexKnauth) [reply types.ss:313]: This has been fixed now, it no longer defines `t` to return it immediately.

File: json-rpc.ss (hunk 1 above)
Comment (AlexKnauth) [reply json-rpc.ss]: Done.

**Pull request merged**